### PR TITLE
[quant][graphmode] Prepack folding for conv2d

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27403 [quant][graphmode] Minor fix and turn off fold_convbn
* #27348 [quant][graphmode] Add e2e test for conv+bn
* #27347 [quant][graphmode][fix] Don't permute in quantized::conv2d pattern
* #27346 [quant][graphmode][fix] insert_prepack_unpack for conv
* #27269 [refactor] Move tests in test_quantizer.py to test_quantization.py
* #27239 [refactor] Factor out insertPrepackUnpackForLinear
* #27238 [refactor] InsertObservers
* #27237 [quant][graphmode][fix] Make inserted child module names unique
* **#27119 [quant][graphmode] Prepack folding for conv2d**

Summary:
att

Test Plan:
python test/test_jit.py 'TestJit.test_fold_prepack'

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D17717636](https://our.internmc.facebook.com/intern/diff/D17717636)

### CODE REVIEW

File: torch/csrc/jit/passes/quantization.cpp [hunk 1]
```diff
@@ -967,12 +994,27 @@ graph(%self, %a_dequant, %w_scale, %w_zero_point, %w_dtype):
...
+        if (!stride || !padding || !dilation) {
+          TORCH_WARN("Non-constant stride/padding/dilation");
+          continue;
+        }
+        set_conv_params(std::vector<IValue>{stride.value(), padding.value(), dilation.value(), groups.value()});
+      }
       set_weight_bias(std::vector<IValue>{IValue(w_quant), IValue(b)});
       // TODO: we need to make sure this name is unique
-      module.register_module("_packed_linear_weight_bias", wrapper);
+      module.register_module("_packed_weight_bias", wrapper_module);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: This will not work if a module has both `linear` and `conv2d` (or several `conv`s).

File: torch/csrc/jit/passes/quantization.cpp [hunk 2]
```diff
@@ -967,12 +994,27 @@ graph(%self, %a_dequant, %w_scale, %w_zero_point, %w_dtype):
...
+        auto stride = toIntList(match_vmap.at(vmap.at("stride")));
+        auto padding = toIntList(match_vmap.at(vmap.at("padding")));
+        auto dilation = toIntList(match_vmap.at(vmap.at("dilation")));
+        auto groups = toIValue(match_vmap.at(vmap.at("groups")));
+        auto set_conv_params = wrapper_module.get_method("set_conv_params");
+        if (!stride || !padding || !dilation) {
+          TORCH_WARN("Non-constant stride/padding/dilation");
+          continue;
+        }
+        set_conv_params(std::vector<IValue>{stride.value(), padding.value(), dilation.value(), groups.value()});
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: Nit: clang-format.

File: torch/csrc/jit/passes/quantization.cpp [hunk 3]
```diff
@@ -938,19 +954,30 @@ void InsertPrepackUnpack(script::Module& module) {
 void FoldPrepackedWeightIntoModule(
     script::Module& module,
     const std::string& method_name,
-    const script::Module& wrapper_module) {
+    const script::Module& linear_params_module,
+    const script::Module& conv_params_module) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: This looks a bit ugly. Are `linear` and `conv` the only ops that might need packing? Should we consider making this more general?

File: torch/csrc/jit/passes/quantization.cpp [hunk 4]
```diff
@@ -981,19 +1023,19 @@ graph(%self, %a_dequant, %w_scale, %w_zero_point, %w_dtype):
...
       // packed_params = wrapper_module._packed_params
       Value* packed_params_from_attr =
           graph->insertGetAttr(packed_params_module, "_packed_params");
       packed_params_val->replaceAllUsesWith(packed_params_from_attr);
 
       // Delete nodes
-      std::vector<Node*> nodes_to_delete = {w_val->node(),
-                                            b_val->node(),
+      std::vector<Node*> nodes_to_delete = {//w_val->node(),
+                                            //b_val->node(),
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: Why was this changed?

**Review by ZolotukhinM — CHANGES_REQUESTED** [review]

File: torch/csrc/jit/passes/quantization.cpp (hunk 1 above)
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp]: yeah see the comment above, I'm planning to do this in a separate PR

File: torch/csrc/jit/passes/quantization.cpp (hunk 3 above)
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp]: these two should be the only cases, and these are workaround, I think we'll have proper support before we have more use cases.

File: torch/csrc/jit/passes/quantization.cpp (hunk 4 above)
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp]: I changed `%weight` and `%bias` to be the input of the pattern, when I try to delete them they have more uses because they are input of the method, so the uses is not empty for them and I can not delete them:
```
graph(%self : ClassType<Conv2d>,
      %input.1 : Tensor,
      %weight.1 : Tensor):
  %27 : float = prim::Constant[value=0.000733255]()
  %28 : int = prim::Constant[value=0]()
  %29 : int = prim::Constant[value=12]()
  %22 : float = prim::Constant[value=0.0783038]()
  %23 : int = prim::Constant[value=65]()
  %24 : int = prim::Constant[value=13]()
  %input.1.quant : Tensor = aten::quantize_per_tensor(%input.1, %22, %23, %24)
  %input.1.dequant : Tensor = aten::dequantize(%input.1.quant)
  %7 : None = prim::Constant() # /home/jerryzh/anaconda2/envs/ve3/lib/python3.6/site-packages/torch/nn/modules/conv.py:339:36
  %8 : int = prim::Constant[value=1]() # /home/jerryzh/anaconda2/envs/ve3/lib/python3.6/site-packages/torch/nn/modules/conv.py:336:46
  %9 : int[] = prim::Constant[value=[2, 2]]()
  %10 : int[] = prim::Constant[value=[3, 3]]()
  %11 : int[] = prim::Constant[value=[1, 1]]()
  %17 : float = prim::Constant[value=0.0275517]()
  %18 : int = prim::Constant[value=65]()
  %36 : ClassType<ConvPackedParams> = prim::GetAttr[name="_packed_weight_bias"](%self)
  %37 : Tensor = prim::GetAttr[name="_packed_params"](%36)
  %32 : Tensor, %33 : Tensor? = quantized::conv_unpack(%37)
  %34 : Tensor = aten::dequantize(%32)
  %35 : Tensor = aten::conv2d(%input.1.dequant, %34, %7, %9, %10, %11, %8)
  %19 : int = prim::Constant[value=13]()
  %12.quant : Tensor = aten::quantize_per_tensor(%35, %17, %18, %19)
  %12.dequant : Tensor = aten::dequantize(%12.quant)
  return (%12.dequant)
```

File: torch/csrc/jit/passes/quantization.cpp [hunk 5]
```diff
@@ -967,34 +997,54 @@ graph(%self, %a_dequant, %w_scale, %w_zero_point, %w_dtype):
...
+      if (is_conv) {
+        module_name_prefix = "_conv_packed_params_module_for_";
+      } else {
+        module_name_prefix = "_linear_packed_params_module_for_";
+      }
       auto w_quant_val = match_vmap.at(vmap.at("w_quant"));
-      auto packed_params_val = match_vmap.at(vmap.at("packed_params"));
+      // unique name for the module based on %w_quant
+      auto module_name =
+          module_name_prefix + std::to_string(w_quant_val->unique());
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: I suggest not depending on `unique()` and rather generate a unique name ourselves, but it's fine if you do it in a follow-up.

**Review by ZolotukhinM — CHANGES_REQUESTED** [review]
Mostly looks good, please find some remarks inline:

File: test/test_jit.py [hunk 6]
```diff
@@ -1410,30 +1406,50 @@ def forward(self, x):
...
+                w_unpacked, b_unpacked = torch.ops.quantized.conv_unpack(packed)
+                r = torch.nn.functional.conv2d(xq.dequantize(), w_unpacked.dequantize(), b_unpacked, stride, padding, dilation, groups)
+                rq = torch.quantize_per_tensor(r, 0.2, 2, torch.quint8)
+                return rq
+
+        for name, M, data in [('linear', QLinear, torch.randn((5, 5), dtype=torch.float)),
+                              ('conv', QConv, torch.randn((1, 3, 24, 24), dtype=torch.float))]:
+            m = torch.jit.script(M())
+            ref_res = get_forward(m._c)(data)
+            fold_prepack(m)
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply test/test_jit.py]: yeah, you mean split the check in 1441 into a different file? but the behavior we want is these weights are packed properly in this context. I think there is still value to put this here

File: torch/csrc/jit/passes/quantization.cpp [hunk 7]
```diff
@@ -638,6 +639,22 @@ graph(%a_dequant, %w, %b, %w_scale, %w_zero_point, %w_dtype, %stride, %padding,
   rewriter.runOnGraph(graph);
 }
 
+c10::optional<IValue> toIntList(Value* v) {
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp]: it does return intlist, do you mean we should add a check in L645 to check it's an IntList?

File: torch/csrc/jit/passes/quantization.cpp [hunk 8]
```diff
@@ -638,6 +639,22 @@ graph(%a_dequant, %w, %b, %w_scale, %w_zero_point, %w_dtype, %stride, %padding,
...
+  auto* n = v->node();
+  if (n->kind() == prim::Constant) {
+    return toIValue(v);
+  } else if (n->kind() == prim::ListConstruct) {
+    auto e0 = toIValue(n->inputs()[0]);
+    auto e1 = toIValue(n->inputs()[1]);
+    if (!e0 || !e1) {
+      return c10::nullopt;
+    }
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp]: because this is conv2d, I can add a check

File: torch/csrc/jit/passes/quantization.cpp (hunk 5 above)
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp]: this is fixed in follow up PR

File: test/test_jit.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ZolotukhinM) [reply test/test_jit.py]: I meant testing the functionality of these classes directly without using our pass (e.g. using the class directly in Python it to pack, then unpack and verifying that the values are the same, maybe also packing and making sure the packed values are accepted by quantized kernels).

File: torch/csrc/jit/passes/quantization.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (ZolotukhinM) [reply torch/csrc/jit/passes/quantization.cpp]: I think the new version is good.

**Review by ZolotukhinM — APPROVED** [review]
Looks good!

File: test/test_jit.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jerryzh168) [reply test/test_jit.py]: I think this is more about checking the result of fold_prepack rather than test the functionality of the PackedParams modules. basically we want to make sure we insert the module and called `set_weight_bias` on correct values.
